### DISCUSSION TITLE
Custom Serializer on Mutation InputGraphType

### DISCUSSION THREAD

**Discussion by davehog** [graphql-dotnet/graphql-dotnet#3753] [Q&A] [answered]
I have a Mutation that saves a fairly complex object, a `SnapshotProductTemplateModel`, one of the object's properties is itself an object, a `SnapshotModel` that has a custom serializer implemented as a `System.Text.Json.JsonConverter<SnapshotModel>`. This SnapshotModel's schema cannot change (even though it is sub-optimal). The issue is that I can't get the custom schema to validate, so running the query in Graphiql, I just get validation errors like this:

`Variable '$snapshotProductSetTemplate.snapshot.assets[0]' is invalid. Unable to parse input as a 'SnapshotAssetModel' type. Did you provide a List or Scalar value accidentally?`

With that in mind, is there a way to register the custom converter so that it validates the input against it? In lieu of that, is there a way to skip the validation of that particular InputType?

**Comment by Shane32** [comment, marked as answer]
No good way.  `DocumentValidator.cs:122` is going to call `GetVariablesValuesAsync` to parse each variable.  To do so, it follows the structure of the graph type that is defined for the variable type, in your situation coercing each of your input object's fields into a dictionary and then passing it to `ParseDictionary` to get converted to `SnapshotModel`.  If you use `JsonConverter` to convert it to a `SnapshotModel` before hitting `GetVariablesValuesAsync`, it won't be able to find the child fields.

One way you could normally handle a situation like this is to define SnapshotModel as a scalar.  Although it won't have any child fields defined for it through GraphQL, you can deserialize the value through a `JsonConverter` and the scalar can pick it up and return it to your field resolver.  But even this isn't going to work out-of-the-box because GraphQL.NET normally deserializes the variables into an `Inputs` object before it even hits the GraphQL.NET pipeline.  What would trigger the `JsonConverter` to execute in the first place?  You'd have to rewrite the `InputsJsonConverter` to detect SnapshotModels and deserialize them correctly, keeping in mind that this happens before the GraphQL request is even parsed, so it will not be know which variables are SnapshotModels and which are not.  And besides, this won't work for you if you cannot change the schema.

I suggest trying to rewrite your JsonConverter code into the ParseDictionary method of the SnapshotModel's InputObjectGraphType.  Is there a reason this does not work for you?

**Reply by davehog** [reply]
@Shane32 Thank you for the quick response; it is very much appreciated on a Friday afternoon!

Yeah, this pretty much is what I was afraid of. I think I may just have to save these objects via web API rather than using the graph, which is fine, just less aesthetic than I would like. Since the deadline is looming, I am going to go down this route; when I get some more time, I will work on rewriting the ParseDictionary method on the SnapshotInputType.

Thank you again!

Dave
